app/src/org/commcare/activities/QueryRequestActivity.java

```diff
@@ -116,56 +132,176 @@ private void buildPromptUI() {
     }
 
     private void buildPromptEntry(LinearLayout promptsLayout, String promptId,
-                                  DisplayUnit displayUnit, boolean isLastPrompt) {
+                                  QueryPrompt queryPrompt, boolean isLastPrompt) {
         Hashtable<String, String> userAnswers =
                 remoteQuerySessionManager.getUserAnswers();
-        promptsLayout.addView(createPromptMedia(displayUnit));
 
-        EditText promptEditText = new EditText(this);
+        View promptView = LayoutInflater.from(this).inflate(R.layout.query_prompt_layout, promptsLayout, false);
+        setLabelText(promptView, queryPrompt.getDisplay());
+        View inputView;
+        String input = queryPrompt.getInput();
+        if (input != null && input.contentEquals(SELECT1)) {
+            inputView = setUpSpinnerView(promptView, queryPrompt, userAnswers);
+        } else {
+            inputView = setUpEditTextView(promptView, queryPrompt, promptId, userAnswers, isLastPrompt);
+        }
+        setUpBarCodeScanButton(promptView, promptId, queryPrompt);
+
+        promptsLayout.addView(promptView);
+        promptsBoxes.put(promptId, inputView);
+    }
+
+    private void setUpBarCodeScanButton(View promptView, String promptId, QueryPrompt queryPrompt) {
+        ImageView barcodeScannerView = promptView.findViewById(R.id.barcode_scanner);
+        barcodeScannerView.setVisibility(isBarcodeEnabled(queryPrompt) ? View.VISIBLE : View.INVISIBLE);
+        barcodeScannerView.setTag(promptId);
+        barcodeScannerView.setOnClickListener(v ->
+                callBarcodeScanIntent((String)v.getTag())
+        );
+    }
+
+    private Spinner setUpSpinnerView(View promptView, QueryPrompt queryPrompt,
+                                     Hashtable<String, String> userAnswers) {
+        Spinner promptSpinner = promptView.findViewById(R.id.prompt_spinner);
+        promptSpinner.setVisibility(View.VISIBLE);
+        promptView.findViewById(R.id.prompt_et).setVisibility(View.GONE);
+
+        promptSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+            @Override
+            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
+                Vector<SelectChoice> choices = queryPrompt.getItemsetBinding().getChoices();
+                SelectChoice selectChoice = choices.get(position);
+                String oldAnswer = userAnswers.get(queryPrompt.getKey());
+                if(oldAnswer == null || !oldAnswer.contentEquals(selectChoice.getValue())) {
```

Comment:
Am I right in assuming that it's this event loop that propagates out of order changes? 

IE: If this spinner view is the last item in the 213 loop, and its value is changed on 208, this selected method is fired, and the loop is redone in case a previous spinner needs to be changed?